src/lib.rs

```diff
@@ -0,0 +1,489 @@
+                    ret = JitTransition::IncrementedHotCount;
+                }
+            }
+        },
+        JitPhase::Tracing(start_loc) => {
+            // If we have traced one iteration of the loop, stop tracing.
+            if loc == start_loc {
+                // XXX stop Intel PT.
+                JitThreadContext::compile_trace(loc);
+                JitThreadContext::phase_transition(JitPhase::Interpreting);
+                ret = JitTransition::FinishedTracing;
+            } else {
+                // Still tracing.
+                ret = JitTransition::ContinuedTracing;
+            }
+        },
+        JitPhase::RunningTrace(_) => {
+            // The JIT control point should not be called from within a trace.
+            unreachable!();
+        },
+    }
+    debug!("JIT Control: thread={:?}: loc={:?}, hot_count={}, trans={:?}",
+           debug_thread_name(), loc, hot_count(loc), ret);
+    ret
+}
+
+/// JIT Transitions.
+///
+/// The JIT control point uses this to communicate what it just did.
+#[derive(Debug, Eq, PartialEq)]
+pub enum JitTransition {
+    IncrementedHotCount,
+    StartedTracing,
+    ContinuedTracing,
+    // AbortedTracing XXX
+    FinishedTracing,
+    ExecutedTrace,
+}
+
+/// A helper to spawn a named thread.
+///
+/// The thread executes the `f` closure, under a thread named `name`.
+fn named_thread<'s, F, N>(name: N, f: F) -> JoinHandle<()>
+               where F: FnOnce(), F: Send + 'static, N: Into<String> {
+    let thr = thread::Builder::new()
+                              .name(name.into())
+                              .spawn(f)
+                              .unwrap();
+    thr
+}
+
+#[cfg(test)]
+mod tests {
+    use ::*;
+    use std::time;
+
+    /// The number of times to repeat multi-threaded tests.
+    const MULTITHREAD_REPS: u8 = 10;
+
+    /// Reset the JIT state.
+    ///
+    /// This is needed to run successive tests within one process, because we use global state and
+    /// thread-locals. To truly reset the JIT, all tests must also ensure that any threads it
+    /// creates are also joined.
+    fn reset_jit() {
+        // Reset shared JIT state.
+        TRACES.write().unwrap().clear();
+        HOT_COUNTS.write().unwrap().clear();
+
+        // Reset per-thread JIT state. Assuming a single thread for now.
+        JIT_THREAD_CX.with(|k| {
+            k.borrow_mut().reset();
+        });
+    }
+
+    /// Test helper to artificially force a trace to be collected for the given location.
+    ///
+    /// The hot counts and JIT transitions are checked after each call to the JIT control point.
+    ///
+    /// The hot count for the desired location must be zero before calling this and you must be
+    /// very careful when using this function in multi-threaded tests. Two threads must not
+    /// concurrently call this function on the same location, otherwise the hot counts and JIT
+    /// transitions cannot be checked due to non-determinism.
```

Comment:
I don't think the bit after "otherwise" is quite right: they *can* be checked, but the values are non-deterministic and thus it's hard to determine if they're correct are not.

Reply:
I think you said an equivalent statement TBH, but I will reword it anyways.

Reply:
"cannot be checked" sounds like "there is no way to read the values". There is a way to read them, but what is read is then hard to make any sense of.